secret_store/src/listener/service_contract_listener.rs

```diff
@@ -0,0 +1,699 @@
+				Self::generate_server_key(&data, &server_key_id, &threshold)
+					.and_then(|server_key| Self::publish_server_key(&data, &server_key_id, &server_key))
+					.map(|_| {
+						trace!(target: "secretstore", "{}: processed GenerateServerKey({}, {}) request",
+							data.self_key_pair.public(), server_key_id, threshold);
+						()
+					})
+					.map_err(|error| {
+						warn!(target: "secretstore", "{}: failed to process GenerateServerKey({}, {}) request with: {}",
+							data.self_key_pair.public(), server_key_id, threshold, error);
+						error
+					})
+			},
+			ServiceTask::Shutdown => unreachable!("it must be checked outside"),
+		}
+	}
+
+	/// Retry processing pending requests.
+	fn retry_pending_requests(data: &Arc<ServiceContractListenerData>) -> Result<usize, String> {
+		let mut failed_requests = 0;
+		let mut processed_requests = 0;
+		let retry_data = ::std::mem::replace(&mut *data.retry_data.lock(), Default::default());
+		for (is_confirmed, task) in data.contract.read_pending_requests() {
+			// only process requests, which we haven't confirmed yet
+			if is_confirmed {
+				continue;
+			}
+
+			let request_result = match task {
+				ServiceTask::GenerateServerKey(server_key_id, threshold) => {
+					// only process request, which haven't been processed recently
+					// there could be a lag when we've just generated server key && retrying on the same block
+					// (or before our tx is mined) - state is not updated yet
+					if retry_data.generated_keys.contains(&server_key_id){
+						continue;
+					}
+
+					// process request
+					let is_own_request = is_processed_by_this_key_server(&*data.key_server_set, &*data.self_key_pair, &server_key_id);
+					Self::process_service_task(data, match is_own_request {
+						true => ServiceTask::GenerateServerKey(server_key_id, threshold.into()),
+						false => ServiceTask::RestoreServerKey(server_key_id),
+					})
+				},
+				_ => Err("not supported".into()),
+			};
+
+			// process request result
+			match request_result {
+				Ok(_) => processed_requests += 1,
+				Err(_) => {
+					failed_requests += 1;
+					if failed_requests > MAX_FAILED_RETRY_REQUESTS {
+						return Err("too many failed requests".into());
+					}
+				},
+			}
+		}
+
+		Ok(processed_requests)
+	}
+
+	/// Generate server key.
+	fn generate_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId, threshold: &H256) -> Result<Public, String> {
+		let threshold_num = threshold.low_u64();
+		if threshold != &threshold_num.into() || threshold_num >= ::std::usize::MAX as u64 {
+			return Err(format!("invalid threshold {:?}", threshold));
+		}
+
+		// key server expects signed server_key_id in server_key_generation procedure
+		// only signer could store document key for this server key later
+		// => this API (server key generation) is not suitable for usage in encryption via contract endpoint
+		let author_key = Random.generate().map_err(|e| format!("{}", e))?;
```

Comment:
`Random.generate()` is infallible.

Reply:
Yes, but it returns a `Result` :) Let's fix the origin of the problem - i.e. `Random.generate()`. Not in this PR, of course. I'll add this to my local txt-file based issue tracker.